RFCs/rfc-0004-fast-transactions.md

```diff
@@ -0,0 +1,415 @@
+
+What is different with SmartContract is that the precondition is checked to include the transaction to the block. This precondition is now checked against the last block account state, so you cannot include several transactions that rely on different precondition states, you should wait for the next block. That is why we need a guaranteed state and a calculated state:
+
+### The guaranteed state
+
+The “guaranteed new state” is very similar to finality, but there are edge cases when they are different, so I prefer to use a different term than finality.
+
+The guaranteed new state is guaranteed to zkApp to become final, so if the architecture of zkApp previews that guaranteed new state should always become final, it is possible, and those two terms are equal.
+
+Still, there is a technical possibility for zkApp to replace the transaction with the same nonce, not using the new state but using the old state, effectively rolling back the transactions not included yet in the Mina blockchain block.
+
+If we formulate what a guaranteed state means in the contract terms, it will lead us to the following wording:
+
+### **Terms of the State Guarantee Contract**
+
+- **Guarantee Issuer**: BlockSpace Node
+
+- **Beneficiary**: zkApp
+
+- **Condition Precedent**: zkApp is obliged at all times during the term of this agreement to maintain at its own expense the fee for the transactions that is necessary for avoiding the eviction of the transaction from the mempool and is enough for the inclusion of the transaction in the block during the maximum time that transaction is allowed to be in the mempool.
+
+- **Term**: The time that starts when zkApp sends the transaction to the BlockSpace Node to be committed to the L1 blockchain and ends on the earliest of:
+
+  - The transaction is included in the L1 block
+  - ZkApp calls off the transaction by sending the replacement transaction
+  - The transaction is excluded from the mempool due to the low fee or due to the maximum time reached for the transaction to be in the mempool as the result of the violations by zkApp of the Condition Precedent.
+
+- **Rights of Third Parties**: Third Parties, including, but not limited to, other zkApps and Accounts, are NOT beneficiaries of the State Guarantee Contract and can rely on this guaranteed state only in the optimistic sense by also relying on the architecture of the zkApp and not on this Guarantee. It is recommended for third parties to wait for BlockSpace to commit to the L1 before executing any material transaction that relies on the state, such as asset withdrawal.
+- **Applicable law**: Mina protocol
+- **Arbitrage**: O(1)Labs
+- **SLA**: TBD
+
```

Comment:
This is very odd, and still doesn't answer any of the essential questions of the guarantee that make it useful. It would be better to delete this and add answers to the following:
* What does the `BlockSpace` node do?
* How do users or the zkApp choose/know which node to go to?
* What are the censorship consequences of this design?
* How does a block producer find out about the `BlockSpace` node's sequenced transactions?
  - Please include specific technical communication mechanisms.
* Why would a block producer include the `BlockSpace` node's sequenced transactions?
  - Technical communication mechanisms needed here too: how does it know that there is a `BlockSpace` node? How does it know there's not more than one racing? How does it know that there are pending transactions, and so it shouldn't execute a one-shot txn for the zkApp? etc.
* Why wouldn't a block producer include some non-`BlockSpace` node's transaction for a zkApp and invalidate all of the work?
  - Specifically, please talk about the incentives. The Mina block producer node would rather maximize their fee; if the `BlockSpace` node's bundle isn't ready when their slot comes around, they can sacrifice the fee to preserve the bundle, but most will just take the fee and ruin all the `BlockSpace` node's work.